src/xkeys.rs

```diff
@@ -0,0 +1,371 @@
+        let raw = decode_raw(source_bytes)?;
+
+        let b1 = raw[0] & 248;
+        if b1 != PREFIX_BYTE_SEED {
+            return Err(err!(
+                InvalidPrefix,
+                "Incorrect byte prefix: {}",
+                source.chars().next().unwrap()
+            ));
+        }
+
+        let b2 = (raw[0] & 7) << 5 | ((raw[1] & 248) >> 3);
+        if b2 != PREFIX_BYTE_CURVE {
+            return Err(err!(
+                InvalidPrefix,
+                "Expect a cruve, got {:?}",
+                KeyPairType::from(b2)
+            ));
+        }
+
+        let mut seed = [0u8; 32];
+        seed.copy_from_slice(&raw[2..]);
+
+        let secret = SecretKey::from_bytes(seed);
+        Ok(Self {
+            public: secret.public_key(),
+            secret: Some(secret),
+        })
+    }
+
+    /// Attempts to return the encoded, human-readable string for this key pair's seed.
+    /// Remember that this value should be treated as a secret. Do not store it for
+    /// any longer than necessary
+    pub fn seed(&self) -> Result<String> {
+        let Some(secret) = &self.secret else {
+            return Err(err!(IncorrectKeyType, "This keypair has no seed"));
+        };
+
+        Ok(encode_seed(&KeyPairType::Curve, &secret.to_bytes()))
+    }
+
+    /// Returns the encoded, human-readable public key of this key pair
+    pub fn public_key(&self) -> String {
+        encode(&KeyPairType::Curve, self.public.as_bytes())
+    }
+
+    pub fn private_key(&self) -> Result<String> {
+        let Some(secret) = &self.secret else {
+            return Err(err!(IncorrectKeyType, "This keypair has no seed"));
+        };
+
+        Ok(encode_prefix(&[PREFIX_BYTE_PRIVATE], &secret.to_bytes()))
+    }
+
+    /// Returns the type of this key pair.
+    pub fn key_pair_type(&self) -> KeyPairType {
+        KeyPairType::Curve
+    }
+
+    pub fn open(&self, input: &[u8], sender: &Self) -> Result<Vec<u8>> {
+        let nonce_size = <SalsaBox as AeadCore>::NonceSize::to_usize();
+
+        let Some(secret_key) = &self.secret else {
+            return Err(err!(SignatureError, "Cannot open without a private key"));
+        };
+
+        if input.len() <= XKEY_VERSION_V1.len() + nonce_size {
+            return Err(err!(InvalidPayload, "Payload too short"));
+        }
+
+        let Some(input) = input.strip_prefix(XKEY_VERSION_V1) else {
+            return Err(err!(InvalidPrefix, "Cannot open message, wrong version"));
+        };
+
+        let (nonce, input) = input.split_at(nonce_size);
+
+        let b = SalsaBox::new(&sender.public, secret_key);
+        b.decrypt(nonce.into(), input)
+            .map_err(|_| err!(InvalidPayload, "Cannot decrypt payload"))
+    }
+
+    /// Seal is compatible with nacl.Box.Seal() and can be used in similar situations for small
+    /// messages. We generate the nonce from crypto rand by default.
+    pub fn seal(&self, input: &[u8], recipient: &Self) -> Result<Vec<u8>> {
```

Comment:
This and `seal_with_rand` will need `#[cfg(not(target_arch = "wasm32"))]` on them as well since they use the random generator

Reply:
Done, I also created `seal_with_nonce` where you provide your own random byte for wasm